File: Sliding Window Pattern/Longest_Substring_With_No_Repeating_Character.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

int findSubString(string s) {
    int windowStart = 0, maxLength = 0;
    unordered_map<char, int> dict;
    for(int windowEnd = 0; windowEnd < s.size(); windowEnd++) {
        char currChar = s[windowEnd];
        if (dict.count(currChar) > 0)
            windowStart = max(windowStart, dict[currChar] + 1);
        dict[currChar] = windowEnd;
        maxLength = max(maxLength, windowEnd - windowStart + 1);
    }
    return maxLength;
}


int main() {
    string s = "abcabcbb";
    cout<<findSubString(s)<<endl;
    return 0;
}

File: Sliding Window Pattern/Longest_Substring_With_No_Repeating_Character.py
def findSubString(s:str):
    windowStart = 0
    maxLength = 0
    dictionary = {}

    for windowEnd in range(len(s)):
        currChar = s[windowEnd]
        if currChar in dictionary:
            windowStart = max(windowStart, dictionary[currChar] + 1)
        dictionary[currChar] = windowEnd
        maxLength = max(maxLength, windowEnd - windowStart + 1)
    return maxLength


if __name__ == "__main__":
    s = "abcabcbb"
    print(findSubString(s))